packages/framework/presence/src/container-definitions/containerExtensions.ts

```diff
@@ -0,0 +1,162 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import type {
+	JsonDeserialized,
+	JsonSerializable,
+} from "@fluid-experimental/presence/internal/core-interfaces";
+
+/**
+ * While connected, the id of a client within a session.
+ *
+ * @internal
+ */
+export type ClientConnectionId = string;
+
+/**
+ * Common interface between incoming and outgoing extension signals.
+ *
+ * @sealed
+ * @internal
+ */
+export interface IExtensionMessage<TType extends string = string, TContent = unknown> {
+	/**
+	 * Message type
+	 */
+	type: TType;
+
+	/**
+	 * Message content
+	 */
+	content: JsonDeserialized<TContent>;
+
+	/**
+	 * The client ID that submitted the message.
+	 * For server generated messages the clientId will be null.
+	 */
+	// eslint-disable-next-line @rushstack/no-new-null
+	clientId: ClientConnectionId | null;
+
+	/**
+	 * Client ID of the singular client the message is being (or has been) sent to.
+	 * May only be specified when IConnect.supportedFeatures['submit_signals_v2'] is true, will throw otherwise.
+	 */
+	targetClientId?: ClientConnectionId;
+}
+
+/**
+ * Defines requirements for a component to register with container as an extension.
+ *
+ * @internal
+ */
+export interface IContainerExtension<TContext extends unknown[]> {
+	/**
+	 * Notifies the extension of a new use context.
+	 *
+	 * @param context - Context new reference to extension is acquired within
+	 */
+	onNewContext(...context: TContext): void;
+
+	/**
+	 * Callback for signal sent by this extension.
+	 *
+	 * @param address - Address of the signal
+	 * @param signal - Signal content and metadata
+	 * @param local - True if signal was sent by this client
+	 */
+	processSignal?(address: string, signal: IExtensionMessage, local: boolean): void;
```

Comment:
small question: noticed that this is optional, would container extensions normally involve dealing with signals(?) or is this more specific to presence extension. if not, maybe this could be removed.

Reply:
The extensions may or may not want to process signals. When this is fully built out there will likely also be an optional processOp(s) member.
I would not look to hard at this portion - it is all unused speculation, but here to keep needs "documented".